### DISCUSSION TITLE
Full duplex CRSF for cheap ELRS

### DISCUSSION THREAD

**Discussion by ggmaniack** [OpenI6X/opentx#322] [Ideas]
Would it be possible to implement Full duplex (separate RX and TX) CRSF communication for ELRS purposes? This would allow using a cheap ESP8 based ELRS receiver with PA as a ~100mW transmitter (or ~10mW without PA). This could more than halve the expense of the ELRS expansion (and would make for easier internal mounting).

Half-duplex inverted ("s.port") input is currently only supported on ELRS receivers which use the Pico MCU/ESP32 (EP1 Dual, SuperD, etc), as part of the Gemini project, but those are a bit more expensive.

If someone could point me to the right files that need changing, I'd happily do the changes myself. I'm currently working on similar changes for the Erfly6 project.

**Comment by ajjjjjjjj** [comment]
Hi, sure!

Take a look at: 
- `radio/src/targets/flysky/telemetry_driver.cpp`  for actual CRSF serial driver.
- `radio/src/targets/flysky/hal.h` for `TELEMETRY_*` entries.

Open question is a way to make it available: as build time option or runtime configuration in hardware settings.

**Comment by jiri-jirus** [comment]
which rx/tx pins would be good for connecting this 'tx as tx' to? would i need to sacrifice bluetooth uart?

**Comment by ajjjjjjjj** [comment]
RX for UART2 used for CRSF is PA15.
The only thing you need to do is change configuration in port initialization to full duplex one like in AUX Serial driver.
